lib/scripting/grammar/grammar.ts

```diff
@@ -0,0 +1,289 @@
+							 * 3) Literal Keyword - For j=1 To 20
+							 * 4) ) Keyword - Function MyFunction(value) End Function
+							 */
+							if (
+								prevToken.type === 'Keyword' ||
+								prevToken.type === 'Identifier' ||
+								prevToken.type === 'Literal' ||
+								prevToken.text === ')'
+							) {
+								tokens.push(' ');
+							}
+						} else if (token.type === 'Identifier') {
+							/**
+							 * Add spaces for the following:
+							 * 1) Keyword Identifier - Sub BallRelease()
+							 * 2) Identifier Identifier - PlaySound SoundFX("fx_flipperup", ...
+							 * 3) ) Identifier - Sub BallRelease_Hit() BallRelease.CreateBall
+							 */
+							if (
+								prevToken.type === 'Keyword' ||
+								prevToken.type === 'Identifier' ||
+								prevToken.text === ')'
+							) {
+								tokens.push(' ');
+							}
+						} else if (token.type === 'Literal') {
+							/**
+							 * Add spaces for the following:
+							 * 1) Keyword Literal - For j=1 To 20
+							 * 2) Identifier Literal - BallRelease 5, -2
+							 */
+							if (prevToken.type === 'Keyword' || prevToken.type === 'Identifier') {
+								tokens.push(' ');
+							}
+						} else if (token.text === '(' || token.text === '-') {
+							/**
+							 * Add spaces for the following:
+							 * 1) Keyword '(' - For ii=(oldSize*2)+1 To(newSize*2):mSlot(ii)=0:Next
+							 * 2) Keyword '-' - For ii=UBound(mSlot) To 0 Step -1:str=str&mSlot(ii):Next
+							 */
+							if (prevToken.type === 'Keyword') {
+								tokens.push(' ');
+							}
+						} else if (token.text === '.') {
+							/**
+							 * Add space for the following:
+							 * 1) <Space>. - case keyReset .Stop
+							 * Do not add a space for the following:
+							 * 1) :<Space>. - Case keyDown swCopy=swDown: .Switch(swCopy)=False
+							 * 2) =<Space>. - If .Exists(aBall) Then .Item(aBall)=.Item(aBall)+1
+							 * 3) +<Space>. - dips(0)=.Dip(0)+.Dip(1)*256+ .Dip(2)*65536+(.Dip(3) And &H7f)*&H1000000
+							 */
+							if (separator && (prevToken.type !== 'Operator' && prevToken.text !== ':')) {
+								tokens.push(' ');
+							}
+						}
+					}
+					tokens.push(token.text);
+					prevToken = token;
+					separator = false;
+				}
+			},
+		});
+
+		return output;
+	}
+
+	public transpile(script: string): Program {
+		const stmts: Statement[] = [];
+
+		const formattedScript = this.format(script);
+
+		const vbsAst = this.parser.getAST(formattedScript, this.GRAMMAR_TARGET_PROGRAM);
+
+		if (vbsAst === null) {
+			throw new Error('Unable to transpile script:\n\n' + formattedScript);
+		}
+
+		const postProcessors = this.postProcessors;
+
+		dashAst(vbsAst, {
+			leave(node: ESIToken, parent: ESIToken) {
+				if (node.type === 'Program') {
+					for (const child of node.children) {
+						if (child.estree) {
+							if (!Array.isArray(child.estree)) {
+								stmts.push(child.estree);
+							} else {
+								stmts.push(...child.estree);
+							}
+						}
+					}
+				} else {
```

Comment:
minor: `return program(stmts);` here and avoid the else block might make it more readable

Reply:
Yes I can do that. Lately I have been struggling with my own person coding practices. 😃

I personally always like returning in one place. To me it makes the code more readable because you know where the return is going to be.

However, as I spend 99% of my day doing Swift, and then see a lot of this, people return everywhere.

Reply:
Personally I either return fast at the beginning of a method to avoid identing, or at the very end, because returning in the middle of the function logic can be easy to overread and thus difficult to debug.